### PR TITLE
Speaker recognition

### PR DESCRIPTION
**Pull request by nithinraok**
**Change Log**

1. Added new Decoder class for speaker Recognition
2. Added StatsPool layer that supports gram, superVector and x-vector 
3. Added new collate function to limit signal sample if its more than x seconds. If this parameter is 0 then it would consider all the batch samples as is
4. Added scripts/get_hi-mia_data.py to download, extract and process manifest files. Also splits dev set + creates utt2spk file (to later support for kaldi PLDA) 
5. Added following scripts to examples/asr : 
     -  speaker_reco.py -> this facilitates training and finetuning of speaker recognition models
     - spkr_get_emb.py -> this is to get embeddings from pretrained checkpoints
     - hi-mia_eval.py -> Scores cosine similarity for hi-mia trial file
     - kaldi_plda.py -> to convert embeddings to kaldi format for PLDA training
     - train_plda.sh -> bash script to train and score PLDA backend scores
6. scripts/scp_to_manifet.py this helps to convert any speaker-related dataset to manifest type

### CODE REVIEW

File: nemo/collections/asr/jasper.py [hunk 1]
```diff
@@ -317,3 +317,179 @@ def forward(self, encoder_output):
...
+
+class JasperDecoderForSpkrClass(TrainableNM):
+    """
+    Jasper Decoder creates the final layer in Jasper that maps from the outputs
+    of Jasper Encoder to the embedding layer followed by speaker based softmax loss.
+
+    Args:
+        feat_in (int): Number of channels being input to this module
+        num_classes (int): Number of unique speakers in dataset
+        emb_sizes (list) : shapes of intermidiate embedding layers (we consider speaker embbeddings from 1st of this layers)
```
Comment (blisc) [inline nemo/collections/asr/jasper.py:330]: intermediate is misspelled.

File: nemo/collections/asr/jasper.py [hunk 2]
```diff
@@ -317,3 +317,179 @@ def forward(self, encoder_output):
...
+        gram (bool) : If we should consider gram based pooling
+        superVector (bool) : If we should unify  x-vector based pooling with gram based pooling
+        init_mode (str): Describes how neural network parameters are
+            initialized. Options are ['xavier_uniform', 'xavier_normal',
+            'kaiming_uniform','kaiming_normal'].
+            Defaults to "xavier_uniform".
```
Comment (blisc) [inline nemo/collections/asr/jasper.py]: Can you add defaults to all args in your docstring?

Comment (blisc) [reply nemo/collections/asr/jasper.py]: Can you remove the gram and super_vector args and just add a mode option with xvector, gram, and super as options?>

File: nemo/collections/asr/jasper.py [hunk 3]
```diff
@@ -317,3 +317,179 @@ def forward(self, encoder_output):
...
+
+        if super_vector and gram:
+            self._feat_in += 2 * feat_in
+
+        self._midEmbd1 = int(emb_sizes[0])  # Spkr Vector Embedding Shape
+        self._midEmbd2 = int(emb_sizes[1]) if len(emb_sizes) > 1 else 0  # Spkr Vector Embedding Shape
+
+        self._num_classes = num_classes
+        self._pooling = StatsPoolLayer(gram=gram, super_vector=super_vector)
+        self.norm = nn.BatchNorm1d(feat_in)
```
Comment (blisc) [inline nemo/collections/asr/jasper.py:389]: Is this used anywhere?

File: nemo/collections/asr/jasper.py [hunk 4]
```diff
@@ -317,3 +317,179 @@ def forward(self, encoder_output):
...
+        pool = self._pooling(encoder_output)
+        mid1, emb1 = self.mid1(pool), self.mid1[:2](pool)
+        mid2, embs = self.mid2(mid1), self.mid2[:2](mid1)
+        out = self.final(mid2)
+
+        return out, emb1
+
+
+# Siamese Network, support to be added in future releases
+class SiameseDecoderForSpeakerClass(TrainableNM):
```
Comment (blisc) [inline nemo/collections/asr/jasper.py:417]: Does this class work? If not, can you comment it out for now?

File: nemo/collections/asr/parts/dataset.py [hunk 5]
```diff
@@ -53,6 +53,56 @@ def seq_collate_fn(batch, token_pad_value=0):
...
 
+def fixed_seq_collate_fn(batch, fixed_length=16000):
+    """collate batch of audio sig, audio len, tokens, tokens len
+
+    Args:
+        batch (Optional[FloatTensor], Optional[LongTensor], LongTensor,
+               LongTensor):  A tuple of tuples of signal, signal lengths,
+               encoded tokens, and encoded tokens length.  This collate func
+               assumes the signals are 1d torch tensors (i.e. mono audio).
+
```
Comment (blisc) [inline nemo/collections/asr/parts/dataset.py:64]: Can you add fixed_length?

File: nemo/collections/asr/parts/dataset.py [hunk 6]
```diff
@@ -376,14 +426,17 @@ def __init__(
         self.trim = trim
         self.load_audio = load_audio
 
-        self.labels = labels
-        self.num_commands = len(labels)
+        self.labels = labels if labels else self.collection.uniq_labels
```
Comment (blisc) [inline nemo/collections/asr/parts/dataset.py:429]: If labels are now optional, can you reflect that in the init() and docstring?

File: nemo/collections/asr/parts/dataset.py [hunk 7]
```diff
@@ -376,14 +426,17 @@ def __init__(
...
+        self.num_commands = len(self.labels)
 
         self.label2id, self.id2label = {}, {}
-        for label_id, label in enumerate(labels):
+        for label_id, label in enumerate(self.labels):
             self.label2id[label] = label_id
             self.id2label[label_id] = label
 
+        for idx in range(len(self.labels[:5])):
+            logging.info(" label id {} and its mapped label {}".format(idx, self.id2label[idx]))
```
Comment (blisc) [inline nemo/collections/asr/parts/dataset.py:438]: should this be info or debug?

File: scripts/get_hi-mia_data.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+import argparse
+import json
+import logging
+import os
+import tarfile
+import urllib.request
+from glob import glob
+
+import librosa as l
+from sklearn.model_selection import StratifiedShuffleSplit
```
Comment (blisc) [inline scripts/get_hi-mia_data.py:26]: do we have sklearn in our requirements?

File: nemo/collections/asr/jasper.py (hunk 1 above)
Comment (nithinraok) [reply nemo/collections/asr/jasper.py]: Noted and updated

File: examples/speaker_recognition/configs/quartznet_spkr_3x2x512_xvector.yaml [hunk 9]
```diff
@@ -0,0 +1,82 @@
+model: "QuartzNet"
```
Comment (blisc) [inline examples/speaker_recognition/configs/quartznet_spkr_3x2x512_xvector.yaml:1]: Please make a new model name?

File: examples/speaker_recognition/configs/quartznet_spkr_5x1x512_xvector.yaml [hunk 10]
```diff
@@ -0,0 +1,100 @@
+model: "QuartzNet"
```
Comment (blisc) [inline examples/speaker_recognition/configs/quartznet_spkr_5x1x512_xvector.yaml:1]: Perhaps a new name?

File: examples/speaker_recognition/hi-mia_eval.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        speaker = line.split('.')[0].split('_')[0]
+        test_list[line] = idx
+
+        if speaker in speaker_list:
+            speaker_list[speaker].append(idx)
+        else:
+            speaker_list[speaker] = [idx]
+
+    emb = int(emb_size)
+    # import ipdb; ipdb.set_trace()
```
Comment (blisc) [inline examples/speaker_recognition/hi-mia_eval.py:47]: Can you remove this line?

File: examples/speaker_recognition/hi-mia_eval.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+import os
+
+import numpy as np
+from scipy.interpolate import interp1d
+from scipy.optimize import brentq
+from sklearn.metrics import roc_curve
+from tqdm import tqdm
+
+
+def get_acc(data_root='./myExps/hi-mia/', emb='', emb_labels='', emb_size=512):
```
Comment (blisc) [inline examples/speaker_recognition/hi-mia_eval.py:25]: can you add a docstring?

File: examples/speaker_recognition/hi-mia_eval.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    # import ipdb; ipdb.set_trace()
+    tmp_file = open(trail_file, 'r').readlines()
+    trail_score = open('trial_score.txt', 'w')
+
+    trial_embs = []
+    keys = []
+    all_scores = []
+    all_keys = []
+
+    for line in tqdm(tmp_file):
```
Comment (blisc) [inline examples/speaker_recognition/hi-mia_eval.py:56]: Can you comment on what this loop is doing?

File: examples/speaker_recognition/hi-mia_eval.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    parser.add_argument("--emb_size", help="Embeddings size", type=int, required=True)
+    args = parser.parse_args()
+    root, emb, emb_labels, emb_size = args.data_root, args.emb, args.emb_labels, args.emb_size
+
+    y_score, y = get_acc(data_root=root, emb=emb, emb_labels=emb_labels, emb_size=emb_size)
+    fpr, tpr, thresholds = roc_curve(y, y_score, pos_label=1)
+
+    eer = brentq(lambda x: 1.0 - x - interp1d(fpr, tpr)(x), 0.0, 1.0)
+    # cmd = ['./compute_scores.sh', task]
+    # subprocess.run(cmd)
```
Comment (blisc) [inline examples/speaker_recognition/hi-mia_eval.py:113]: please remove comments

File: examples/speaker_recognition/speaker_reco.py [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+    emb_size = 1024
+    name = construct_name(
+        args.exp_name, args.lr, args.batch_size, args.num_epochs, args.weight_decay, args.optimizer, emb_size=emb_size,
+    )
+    work_dir = name
+    if args.work_dir:
+        work_dir = os.path.join(args.work_dir, name)
+
+    # data_dir = '/data/samsungSSD/NVIDIA/datasets/LibriSpeech/'
+    # abs_dir=os.path.abspath(data_dir)
```
Comment (blisc) [inline examples/speaker_recognition/speaker_reco.py:251]: remove

File: examples/speaker_recognition/spkr_get_emb.py [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+
+    decoder = nemo_asr.JasperDecoderForSpkrClass(
+        feat_in=spkr_params['JasperEncoder']['jasper'][-1]['filters'],
+        num_classes=254,
+        emb_sizes=spkr_params['JasperDecoderForSpkrClass']['emb_sizes'].split(','),
+        gram=spkr_params['JasperDecoderForSpkrClass']['gram'],
+        super_vector=spkr_params['JasperDecoderForSpkrClass']['super_vector'],
+    )
+
+    # weight = pickle.load(open('myExps/all_LibriSpeech/weight.pkl','rb'))
```
Comment (blisc) [inline examples/speaker_recognition/spkr_get_emb.py:132]: remove

File: examples/speaker_recognition/spkr_get_emb.py [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+
+    name = construct_name(
+        args.exp_name, args.lr, args.batch_size, args.num_epochs, args.weight_decay, args.optimizer, args.emb_size
+    )
+    work_dir = name
+    if args.work_dir:
+        work_dir = os.path.join(args.work_dir, name)
+
+    # data_dir = '/data/samsungSSD/NVIDIA/datasets/LibriSpeech/'
+    # abs_dir=os.path.abspath(data_dir)
```
Comment (blisc) [inline examples/speaker_recognition/spkr_get_emb.py:163]: remove

File: scripts/get_hi-mia_data.py (hunk 8 above)
Comment (nithinraok) [reply scripts/get_hi-mia_data.py:26]: sklearn gets installed from one of our dependencies not sure which one

File: nemo/collections/asr/jasper.py (hunk 2 above)
Comment (nithinraok) [reply nemo/collections/asr/jasper.py]: Clean thought. Yes, added pool_mode instead of giving too many options for user 👍🏻

File: nemo/collections/asr/jasper.py (hunk 3 above)
Comment (nithinraok) [reply nemo/collections/asr/jasper.py]: no. removed it

File: nemo/collections/asr/parts/dataset.py (hunk 5 above)
Comment (nithinraok) [reply nemo/collections/asr/parts/dataset.py]: Done

File: nemo/collections/asr/parts/dataset.py (hunk 6 above)
Comment (nithinraok) [reply nemo/collections/asr/parts/dataset.py:437]: Done

File: nemo/collections/asr/parts/dataset.py (hunk 7 above)
Comment (nithinraok) [reply nemo/collections/asr/parts/dataset.py]: changed to debug

File: examples/speaker_recognition/configs/quartznet_spkr_3x2x512_xvector.yaml (hunk 9 above)
Comment (nithinraok) [reply examples/speaker_recognition/configs/quartznet_spkr_3x2x512_xvector.yaml]: Done

File: examples/speaker_recognition/configs/quartznet_spkr_5x1x512_xvector.yaml (hunk 10 above)
Comment (nithinraok) [reply examples/speaker_recognition/configs/quartznet_spkr_5x1x512_xvector.yaml]: Done

File: examples/speaker_recognition/hi-mia_eval.py (hunk 11 above)
Comment (nithinraok) [reply examples/speaker_recognition/hi-mia_eval.py]: Done

File: examples/speaker_recognition/hi-mia_eval.py (hunk 13 above)
Comment (nithinraok) [reply examples/speaker_recognition/hi-mia_eval.py:57]: Done

File: examples/speaker_recognition/hi-mia_eval.py (hunk 14 above)
Comment (nithinraok) [reply examples/speaker_recognition/hi-mia_eval.py]: Done

File: examples/speaker_recognition/speaker_reco.py (hunk 15 above)
Comment (nithinraok) [reply examples/speaker_recognition/speaker_reco.py]: Done

File: examples/speaker_recognition/spkr_get_emb.py (hunk 16 above)
Comment (nithinraok) [reply examples/speaker_recognition/spkr_get_emb.py]: Done

File: examples/speaker_recognition/spkr_get_emb.py (hunk 17 above)
Comment (nithinraok) [reply examples/speaker_recognition/spkr_get_emb.py]: done

Comment (nithinraok) [reply examples/speaker_recognition/spkr_get_emb.py]: done

File: examples/speaker_recognition/spkr_get_emb.py (hunk 16 above)
Comment (nithinraok) [reply examples/speaker_recognition/spkr_get_emb.py]: done

File: examples/speaker_recognition/speaker_reco.py (hunk 15 above)
Comment (nithinraok) [reply examples/speaker_recognition/speaker_reco.py]: done

File: examples/speaker_recognition/hi-mia_eval.py (hunk 14 above)
Comment (nithinraok) [reply examples/speaker_recognition/hi-mia_eval.py]: done

File: examples/speaker_recognition/hi-mia_eval.py (hunk 13 above)
Comment (nithinraok) [reply examples/speaker_recognition/hi-mia_eval.py:57]: done

File: examples/speaker_recognition/hi-mia_eval.py (hunk 11 above)
Comment (nithinraok) [reply examples/speaker_recognition/hi-mia_eval.py]: done

File: examples/speaker_recognition/hi-mia_eval.py (hunk 12 above)
Comment (nithinraok) [reply examples/speaker_recognition/hi-mia_eval.py:38]: done

File: nemo/collections/asr/jasper.py (hunk 4 above)
Comment (nithinraok) [reply nemo/collections/asr/jasper.py]: okay

File: examples/speaker_recognition/hi-mia_eval.py (hunk 12 above)
Comment (nithinraok) [reply examples/speaker_recognition/hi-mia_eval.py:38]: done

File: scripts/get_hi-mia_data.py (hunk 8 above)
Comment (blisc) [reply scripts/get_hi-mia_data.py:26]: Can you add it to requirements_asr.txt to be sure?

Comment (blisc) [reply scripts/get_hi-mia_data.py:26]: Actually it's in requirements_nlp, can you move it to requirements

Comment (nithinraok) [reply scripts/get_hi-mia_data.py:26]: Done
